include/zephyr/drivers/api.h

```diff
@@ -0,0 +1,41 @@
+/**
+ * @file
+ * @brief Generic public API header file.
+ */
+
+/*
+ * Copyright (c) 2024 Basalte bv
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_API_H_
+#define ZEPHYR_INCLUDE_DRIVERS_API_H_
+
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/iterable_sections.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @cond INTERNAL_HIDDEN */
+#define DRIVER_API(_type, _name) const STRUCT_SECTION_ITERABLE(_type, _name)
+
+#define DRIVER_API_IS(_type, _api)                                                                 \
```

Comment:
would it make sense to have an alias for this that takes dev as an argument? It would probably be the common use case for application code, something like

```c
#define DEVICE_API_IS(_type, _dev) DRIVER_API_IS(_type, _dev->api)
```

Reply:
Come to think of it, I think it is more useful to have the inline helper function

```c
static inline bool device_has_adc_driver_api(const struct device *dev)
{
	return COND_CODE_1(CONFIG_ADC, (DRIVER_API_IS(adc_driver_api, dev->api)), (false));
}
```

This allows to call `device_has_adc_driver_api` even if `CONFIG_ADC` is disabled, otherwise we get linker errors.

Reply:
Very cool, I think a potential followup would be to think if we can figure out a way of enumerating and generating these automatically at some intermediate build phase, but frankly I think this may be good enough as is to be presented at the architecture wg, always time to make it more complex, maybe it won't be a problem after all.

I'll tag it for Arch WG if you don't mind, there's probably a queue to get there anyway.